File: src/main/scala/com/mikemunhall/katas/scala/Factorial.scala
package com.mikemunhall.katas.scala

object Factorial {

  // Factorial e.g.: 5! == 5 * 4 * 3 * 2 * 1
  //           e.g.: 0! == 1 (a special case)
  def getResult(x: Int): Int = {
    x match {
      case x if x < 0 => throw new IllegalArgumentException
      case x if x == 0 => 1
      case x => getResult(x - 1) * x
    }
  }

}

File: src/main/java/com/mikemunhall/katas/java/Factorial.java
package com.mikemunhall.katas.java;

public class Factorial {

    public static int getResult(int x) {
        if (x < 0) {
            throw new IllegalArgumentException();
        } else if (x == 0) {
            return 1;
        } else {
            return x * getResult(x - 1);
        }
    }
}
